src/modulus_convert.rs

```diff
@@ -0,0 +1,148 @@
+use crate::error::BoxError;
+use crate::field::Field;
+use crate::helpers::ring::Ring;
+use crate::prss::Participant;
+use crate::replicated_secret_sharing::ReplicatedSecretSharing;
+use crate::securemul::{ProtocolContext, SecureMul};
+use bit_vec::BitVec;
+
+pub enum HelperIdentity {
+    H1,
+    H2,
+    H3,
+}
+
+pub struct RandomShareGenerationHelper<'a> {
+    rng: &'a Participant,
+    identity: HelperIdentity,
+}
+
+impl<'a> RandomShareGenerationHelper<'a> {
+    #[must_use]
+    pub fn new(rng: &'a Participant, identity: HelperIdentity) -> Self {
+        Self { rng, identity }
+    }
+
+    #[must_use]
+    pub fn split_binary<T: Field>(
+        &self,
+        left: bool,
+        right: bool,
+    ) -> (
+        ReplicatedSecretSharing<T>,
+        ReplicatedSecretSharing<T>,
+        ReplicatedSecretSharing<T>,
+    ) {
+        let left = u128::from(left);
+        let right = u128::from(right);
+        match self.identity {
+            HelperIdentity::H1 => (
+                ReplicatedSecretSharing::new(T::from(left), T::ZERO),
+                ReplicatedSecretSharing::new(T::ZERO, T::from(right)),
+                ReplicatedSecretSharing::new(T::ZERO, T::ZERO),
+            ),
+            HelperIdentity::H2 => (
+                ReplicatedSecretSharing::new(T::ZERO, T::ZERO),
+                ReplicatedSecretSharing::new(T::from(left), T::ZERO),
+                ReplicatedSecretSharing::new(T::ZERO, T::from(right)),
+            ),
+            HelperIdentity::H3 => (
+                ReplicatedSecretSharing::new(T::ZERO, T::from(right)),
+                ReplicatedSecretSharing::new(T::ZERO, T::ZERO),
+                ReplicatedSecretSharing::new(T::from(left), T::ZERO),
+            ),
+        }
+    }
+
+    /// Runs the protocol to generate a pair of secret sharings of a random value "r" in [0, 1]
+    /// None of the three helpers will know what the value of "r" is
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, as this method calls the "Secure Mult" implementation
+    /// If there is a timeout (or other error) it should be signalled in the response
+    pub async fn gen_r_pairs<T: Field, R: Ring>(
+        &self,
+        idx: u128,
+        ctx: &ProtocolContext<'_, R>,
+    ) -> Result<Vec<(bool, ReplicatedSecretSharing<T>)>, BoxError> {
+        let mut idx = idx;
+
+        let (left, right) = self.rng.generate_values(idx);
+        let mut output = Vec::with_capacity(128);
+        let left_bits = BitVec::from_bytes(&left.to_be_bytes());
+        let right_bits = BitVec::from_bytes(&right.to_be_bytes());
+        for i in 0..128 {
```

Comment:
I am not confident in the parallelisation here. 

Ideally, I would like all 128 multiplications to happen in parallel. I'm worried that the way I've written it - they'll be run in sequence. I would appreciate reviewers help in figuring out how to ensure there's just a single batch of communication for all of them together.